### ISSUE TITLE
push notifications does not seem to work automatically on every Android device

### ISSUE THREAD

**Issue by petervojtek** [PowerlineApp/powerline-mobile#241]
I have following devices:
- android 4.4 emulator
- android 5 emulator
- android 6 smartphone

and push notifications arrive only to the 4.4 emulator

I tried to login as several different users, but none of them received push notifications on the other devices I use.

**Comment by petervojtek** [comment]
@jterps08 , if you have some other smartphone which was never used with Powerline app, can you try if you will receive the push notifications there?

**Comment by jterps08** [comment]
@petervojtek  I will need to leverage an iPhone of a friend unfortunately... Austin may be able to help us. I will try to get UDID for IPA build for iOS for testing...

**Comment by petervojtek** [comment]
so after I logged out & logged in powerline app on the android 4.4 emulator (which was the only device where push notifications worked for me) , I no longer receive there push notification. 
I also tested all the other Android devices I have and I don't receive push notifications to any of them.

@igaponov, @austinpapp  , is it possible to see some push notifications server logs, to see the actual distribution of the push notifications? I.e. i would like to see something like this:

```
2016-09-20 10:00:09 -- device ID:12345 registered to receive push notifications. Assigned  registrationId:ABCD.
2016-09-20 10:00:20 -- push notification sent to device ID:12345 with content: {some-json}
```

**If you can grant me access to the  Google Play Developer Console account, i should be able to see the logs there:**
https://developers.google.com/cloud-messaging/diagnostics

when I launch the powerline mobile app on android, I successfully register to the push notifications, e.g. I receive this `registrationId:"djC2hiJJh3E:APA91bFk_4WQrYO2SeQbpnupOMffWR2BLaXSGWCaKAd466L1l-fM_YbdRqZWytw6WvrKd7PTs3aCOedBNheULFkqxij13ka4iUqwVO9D8Hm6qUXpyxNLlRVQ4eVPCVvQu8ku-LiFqjYl"`

but after then I receive nothing else.

**Comment by jterps08** [comment]
Access granted to developer console @petervojtek . 

It is _possible_ that the backend is utilizing old GCM instead of new FCM and that is related to this. I don't know.

**Comment by austinpapp** [comment]
i'm also curious about rabbitmq. in the past we have ad serialization problems. idk if this is present. we can check supervisord logs to confirm they are clean.

**Comment by petervojtek** [comment]
ok.. I did some research while having access to the google console and google play console. So far I am unable to access the push messages diagnostics (i.e. to see particular push messages). 

If I understand properly, powerline app is using the legacy GCM to send push notifications. 
There was a screen in the past in Google Play Service Console named "GCM Diagnostics", which allowed to diagnose the messages, but it seems the screen [is no longer available](https://play.google.com/apps/publish/?account=6818514866191666760&dev_acc=00316418262917283982#MarketListingPlace:p=ne.powerli). To confirm my assumption I also [asked question on stackoverflow](http://stackoverflow.com/questions/39619122/push-notification-diagnostics-is-no-longer-available-from-google-play-developer).

I can see the statistics on Google Play Service Console under Reports > Statistics > Powerline > Latest Monthly FCM report, which contains following lines:

```
Date,Package Name,GCM Sender ID,GCM Messages,GCM Registrations
2016-09-01,ne.powerli,354736666363,0,2
2016-09-02,ne.powerli,354736666363,0,1
2016-09-02,ne.powerli,0,0,2
2016-09-03,ne.powerli,354736666363,0,1
2016-09-03,ne.powerli,0,0,2
2016-09-05,ne.powerli,354736666363,0,3
2016-09-06,ne.powerli,354736666363,0,3
2016-09-07,ne.powerli,0,0,1
2016-09-07,ne.powerli,354736666363,0,1
2016-09-08,ne.powerli,0,0,1
2016-09-08,ne.powerli,354736666363,2,1
2016-09-09,ne.powerli,0,0,1
2016-09-09,ne.powerli,354736666363,31,3
2016-09-10,ne.powerli,354736666363,0,5
2016-09-10,ne.powerli,0,0,2
2016-09-11,ne.powerli,0,0,1
2016-09-11,ne.powerli,354736666363,25,3
2016-09-12,ne.powerli,354736666363,63,2
2016-09-13,ne.powerli,354736666363,74,1
2016-09-13,ne.powerli,0,0,2
2016-09-14,ne.powerli,354736666363,18,0
2016-09-15,ne.powerli,354736666363,27,1
2016-09-15,ne.powerli,0,0,1
2016-09-16,ne.powerli,354736666363,31,2
2016-09-16,ne.powerli,0,0,1
2016-09-17,ne.powerli,354736666363,13,0
2016-09-18,ne.powerli,354736666363,22,0
```

The log ends on 2016-09-18, but that is probably because they have some delay on collecting the data (also other reports miss last three days).

There is also some report screen on Google Developer Console:
https://console.cloud.google.com/apis/api/googlecloudmessaging-json.googleapis.com/overview?project=artful-abode-487&duration=P30D

but the stats there are empty, which indicates that no push messages travel via that API.

**To sum up:**
it seems to me the only way how to  be able to diagnose the push messages in detail is [to switch from GCN to FCN](https://support.google.com/googleplay/android-developer/answer/2663268). @austinpapp, I would appreciate if you could confirm my findings -- ideally if you will find out that I am wrong and there still exist the Google Play Develop Console screen where I can find all the push messages and diagnose them.

**Comment by jterps08** [comment]
I do not see stat screen either, 

This may be a relevant document too: https://developers.google.com/cloud-messaging/android/android-migrate-fcm
And here is the Firebase Console: https://console.firebase.google.com/project/artful-abode-487/overview

**Comment by jterps08** [comment]
It seems that neither I or @austinpapp could find anything useful in looking at GCM or SNS. 

What is our next step here @petervojtek @igaponov ?

**Comment by petervojtek** [comment]
Unless I will be able to query push notifications logs, I don't have other options to move with this issue.

**Comment by jterps08** [comment]
Per @austinpapp, @petervojtek please talk to @igaponov about logging into dev server to view logs.

**Comment by petervojtek** [comment]
@igaponov , I observe following behaviour:

I have two android devices:
- android 4.4, logged user Pavlik Morozov (peter.vojtek) -- here the push notifications DO arrive
- android 6, logged user `Peter12` -- here the push notifications are NOT arriving

Pavlik (android 4.4), created Post 295 and is subscribed to it.
Peter12 (android 6) created Post 296 and is subscribed to it.

Now as peter10 I am about to add comments to both of the Posts and I expect push notification to appear in `/dev/shm/powerline/logs/push_dev.log`.

#### Behaviour on the working Android 4.4

When I add comment to Post 295, I see following entry in log and the push notification arrives to the device android 4.4:

```
[2016-09-23 04:46:53] push.DEBUG: Message is pushed {"title":"Peter10 Is Leader","message":" commented on your post","type":"own-post-commented","entityData":{"id":1590,"target":{"id":295,"preview":"pavlik comment 1","type":"post","label":"post","comment_id":613}},"image":"http://powerline-dev.imgix.net/577cb68e6f276.jpeg?dpr=0.75&h=400&ixlib=php-1.1.0&w=320","arn":"arn:aws:sns:eu-west-1:863632456175:endpoint/GCM/powerline_android_dev/8024c99b-dec9-37b9-9b07-377f943fcb9f","user":{"id":78,"username":"peter.vojtek"}} []
```

#### Behaviour on the not-working Android 6

However when adding comments to  Post 296, I observe following behaviour:

I have seen one push notification being generated in log when I added first comment:

```
[2016-09-23 04:44:02] push.DEBUG: Message is pushed {"title":"Peter10 Is Leader","message":" commented on your post","type":"own-post-commented","entityData":{"id":1584,"target":{"id":296,"preview":"hi there!","type":"post","label":"post","comment_id":611}},"image":"http://powerline-dev.imgix.net/577cb68e6f276.jpeg?dpr=0.75&h=400&ixlib=php-1.1.0&w=320","arn":"arn:aws:sns:eu-west-1:863632456175:endpoint/GCM/powerline_android_dev/4e1b40e9-e71f-330b-864a-8b7a8850246c","user":{"id":75,"username":"Peter12"}} []
```

this push notification did not arrived to the Android 6 phone.

When I added additional comment to Post 296, for these comment there was no push notification being generated by the server (no new entry in `/dev/shm/powerline/logs/push_dev.log`).

To sum up -- I added three comments to Post 296, and only the first comment triggered push notification to be generated on server (and even this one did not arrived to smartphone).

It seems server is writing to the log file instantly (instant flush after write), so it is not the case that the logfile i just missing some latest data due to write buffering.

**Comment by austinpapp** [comment]
this arn is wrong `arn:aws:sns:eu-west-1:863632456175:endpoint/GCM/powerline_android_dev/4e1b40e9-xxxxxxxx-8b7a8850246c`

it should be `us-east-1`

i fixed this for dev deployment data and is latest commit to master.

**Comment by austinpapp** [comment]
fixed it on dev server. next deployment shouldn't be a problem for this. 

@petervojtek try to make some updates again please. let me know

**Comment by jterps08** [comment]
Why would it work for my device then? Carryover from old testing efforts
perhaps?

**Comment by austinpapp** [comment]
Depends on when you received the notification. We can chat more offline if you need.

**Comment by petervojtek** [comment]
@igaponov, @austinpapp , I created Post 307 and added comment to it by other user, but push notification related to it have not appeared in `/dev/shm/powerline/logs/push_dev.log`.

**Comment by igaponov** [comment]
@petervojtek user with id=85 (post#307 creator)  doesn't have an endpoint to push.

**Comment by petervojtek** [comment]
@igaponov , what does it means "doesn't have an endpoint to push." ?

the user with ID=85 is logged in on my Android 6 smartphone (the device when no single push notification arrived yet), and when I launch the mobile app on the device, I can see that the push notification registration is successful:

```
succesfully registered to push notifications with ID: ftQLENUtIMo:APA91bFfnXUgLJYtmZejxo4dtU5urwcvcPx7vI5EbfaBCh3WRStXu8_2NhyAS8JfITznf9d3a5FGuYVedYqT6pKvCZEMI3VOCT8So16ViI60B1_f89-J7Cp7neyVCtzb6WGfMwjQzVH_ using senderID: 354736666363
```

**Comment by igaponov** [comment]
@petervojtek 

```
$ http get https://api-dev.powerli.ne/api/endpoints/ "Token: $TOKEN"
HTTP/1.1 200 OK
Access-Control-Allow-Headers: accept, origin, x-requested-with, Authorization, content-type, Token
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS, PATCH
Access-Control-Allow-Origin: *
Cache-Control: no-cache
Connection: keep-alive
Content-Type: application/json
Date: Tue, 27 Sep 2016 17:11:17 GMT
Server: nginx/1.4.6 (Ubuntu)
Transfer-Encoding: chunked
X-Debug-Token: 148f15

[]
```

This user has no endpoint
Use `POST https://api-dev.powerli.ne/api/endpoints/` to add it

**Comment by petervojtek** [comment]
@igaponov , but I don't know what does the word "endpoint" means :( and what should I do to have an endpoint.

**Comment by jterps08** [comment]
Let's try to look at this from a step-by-step standpoint.

User downloads mobile app onto mobile device.
User installs app.
User registers or logs into the app
Something happens with backend server to record that user's endpoint into AWS
Something happens with AWS SNS and GCM/FCM and APNS
Some event triggers push notification on backend server which gets sent to SNS and GCM/APNS?
Unclear about RabbitMQ

Obviously, those steps are probably incorrect, but it would be helpful to use  a step-by-step process to help communicate where we are supposed to use "https://api-dev.powerli.ne/api/endpoints/" in the process.... @igaponov @petervojtek @austinpapp

**Comment by igaponov** [comment]
@petervojtek you should register your app on backend to get push notifications. Use `POST https://api-dev.powerli.ne/api/endpoints/` to add your device. You can check if it's already added with `GET https://api-dev.powerli.ne/api/endpoints/`.

**Comment by jterps08** [comment]
@austinpapp please take a look at this...

I was under the impression that the https://github.com/phonegap/phonegap-plugin-push took care of the endpoint registration, not sure what's going on here, but we need to get past this...

**Comment by petervojtek** [comment]
@igaponov , here is my workflow:

I make this request:

```
Request URL:https://api-dev.powerli.ne/api/endpoints/
Request Method:POST
Status Code:201 Created

{"type":"android","token":"ftQLENUtIMo:APA91bFfnXUgLJYtmZejxo4dtU5urwcvcPx7vI5EbfaBCh3WRStXu8_2NhyAS8JfITznf9d3a5FGuYVedYqT6pKvCZEMI3VOCT8So16ViI60B1_f89-J7Cp7neyVCtzb6WGfMwjQzVH_"}

response:

{"token":"ftQLENUtIMo:APA91bFfnXUgLJYtmZejxo4dtU5urwcvcPx7vI5EbfaBCh3WRStXu8_2NhyAS8JfITznf9d3a5FGuYVedYqT6pKvCZEMI3VOCT8So16ViI60B1_f89-J7Cp7neyVCtzb6WGfMwjQzVH_"}
```

Then I restart the app, and the push notification plugin tells me that this is my token:

```
ftQLENUtIMo:APA91bFfnXUgLJYtmZejxo4dtU5urwcvcPx7vI5EbfaBCh3WRStXu8_2NhyAS8JfITznf9d3a5FGuYVedYqT6pKvCZEMI3VOCT8So16ViI60B1_f89-J7Cp7neyVCtzb6WGfMwjQzVH_
```

so it is the same token as registered above.

However when I make this request:

```
Request URL:https://api-dev.powerli.ne/api/endpoints/
Request Method:GET
Status Code:200 OK
```

I receive empty response:

```
[]
```

**Comment by igaponov** [comment]
@austinpapp `Invalid parameter: PlatformApplicationArn Reason: Region from ARN (us-east-1) is different than SNS endpoint region (eu-west-1).`

**Comment by austinpapp** [comment]
@igaponov where is that error coming from? here is what i see on sns: `
arn:aws:sns:us-east-1:863632456175:app/GCM/powerline_android_dev`

so my question for you is where is eu-west-1 coming from? i looked in the codebase and dont see it anywhere that we would expect it.

**Comment by austinpapp** [comment]
gcm/fcm and apns function exactly the same. generally:
1. mobile client requests token from mobile provider ( say gcm )
2. token is returned from provider
3. mobile client should then hits our http endpoint to store that token. 
   1. igor made reference to that http endpoint above 
   2. this should be working as i noticed tokens being stored. however, there is the ability to delete tokens so i'm not sure of how the workflow exists in mobile side. **we need to understand that part more clearly**
4. our API stores this token and makes an update to AWS SNS ARN referenced in the config file with that token forming a complete ARN
5. _push notification worthy action happens_
6. our API event listener picks up the social activity and drops the event on rabbitmq (our messaging middleware for async action)
7. we have a consumer listening on the end of the message queue (PushSender i think the name is) that will take the event and task AWS SNS to send a notification to the complete ARN. 
8. AWS SNS talks to our expected platform (GCM) that has that token registered ( resolvable via the ARN)
9. _notification is delivered_

**Comment by petervojtek** [comment]
@austinpapp , regarding 3 ii:

the mobile app workflow is following:
1. register to push notifications plugin, obtain registrationID (=token)
2. `GET /api/endpoints/`
3. if the token from step 1 is present in response from step 2, we are done.
4. if the token from step 1 is NOT present in response from step 2, make `POST /api/endpoints/` with the token

UPDATE:
I searched the mobile app source code and there isn't any place where `DELETE /api/endpoints` could be called

**Comment by jterps08** [comment]
Very interesting... I appreciate the clarity.

**Comment by igaponov** [comment]
@austinpapp the problem was in region of an AWS client, I've changed it in configs, should work now.

**Comment by petervojtek** [comment]
@austinpapp , @igaponov , now I properly receive my token:

```
Request URL:https://api-dev.powerli.ne/api/endpoints/
Request Method:GET
Status Code:200 OK

[{"id":279,"token":"ftQLENUtIMo:APA91bFfnXUgLJYtmZejxo4dtU5urwcvcPx7vI5EbfaBCh3WRStXu8_2NhyAS8JfITznf9d3a5FGuYVedYqT6pKvCZEMI3VOCT8So16ViI60B1_f89-J7Cp7neyVCtzb6WGfMwjQzVH_"}]
```

.When other user comments on my post, I receive no push notifications to the device, but I see following entries in`/dev/shm/powerline/logs/push_dev.log` 

```
[2016-09-28 13:13:44] push.DEBUG: Message is pushed {"title":"Peter31 Peter31","message":" commented on your post","type":"own-post-commented","entityData":{"id":1825,"target":{"id":312,"preview":"comment 1","type":"post","label":"post","comment_id":649}},"image":"https://api-dev.powerli.ne/bundles/civixfront/img/logo_320x320.jpg","token":"ftQLENUtIMo:APA91bFfnXUgLJYtmZejxo4dtU5urwcvcPx7vI5EbfaBCh3WRStXu8_2NhyAS8JfITznf9d3a5FGuYVedYqT6pKvCZEMI3VOCT8So16ViI60B1_f89-J7Cp7neyVCtzb6WGfMwjQzVH_","arn":"arn:aws:sns:us-east-1:863632456175:endpoint/GCM/powerline_android_dev/4e1b40e9-e71f-330b-864a-8b7a8850246c","user":{"id":85,"username":"test2"}} []


[2016-09-28 13:15:07] push.DEBUG: Endpoint is disabled and will be removed {"token":"ftQLENUtIMo:APA91bFfnXUgLJYtmZejxo4dtU5urwcvcPx7vI5EbfaBCh3WRStXu8_2NhyAS8JfITznf9d3a5FGuYVedYqT6pKvCZEMI3VOCT8So16ViI60B1_f89-J7Cp7neyVCtzb6WGfMwjQzVH_","arn":"arn:aws:sns:us-east-1:863632456175:endpoint/GCM/powerline_android_dev/4e1b40e9-e71f-330b-864a-8b7a8850246c","user":{"id":85,"username":"test2"}} []
```

**Comment by igaponov** [comment]
@petervojtek I get this error from SNS. 
http://stackoverflow.com/questions/19374411/getting-endpointdisabled-from-amazon-sns

**Comment by petervojtek** [comment]
@igaponov , I experience this problems only on the real android device. It is possible that in the past I installed the old production app on this device (from play store) -- given the stackoverflow link above, this may explain why we see the `Endpoint is disabled` behavior..

**Comment by jterps08** [comment]
I always uninstalled production app from my device before installing test build... don't know if that's relevant or not but wanted to share

**Comment by petervojtek** [comment]
@jterps08 , that is what I did prior I wrote the previous comment -- but the `Endpoint is disabled and will be removed` emerged despite that.  Actually, when you have production powerline app installed and try to install testing powerline app (and have not uninstalled the production app before), you won't be allowed to do that by android OS due to app signing key mismatch.

**Comment by jterps08** [comment]
@austinpapp what do we need here?

**Comment by petervojtek** [comment]
ok, here is the solution:

If you experience on android that you do not receive push notifications, go to android Settings > Apps > Powerline > and choose "Clear all app data" and "Clear cache".

this issue will occur when you use production Powerline app and then switch to development Powerline app, or vice versa.

To sum, proper steps to switch from production to development Powerline app (or vice versa):
1. Uninstall production app
2. Install development app
3. choose Android > Settings > Apps > Powerline > and choose "Clear all app data" and "Clear cache".

**Comment by jterps08** [comment]
NO way... haha okay. That works for me.
